libraries/package-deps-hash/src/getRepoState.ts

```diff
@@ -246,6 +246,109 @@ export function applyWorkingTreeState(
   }
 }
 
+/**
+ * Get the list of all git submodules path
+ * @param rootDirectory - The root directory of the Git repository
+ * @param gitPath - The path to the Git executable
+ * @internal
+ */
+export function getSubmodulePaths(rootDirectory: string, gitPath?: string): string[] {
+  const submodulePaths: string[] = [];
+
+  const submoduleStatusResult: child_process.SpawnSyncReturns<string> = Executable.spawnSync(
+    gitPath || 'git',
+    ['submodule', 'status'],
+    {
+      currentWorkingDirectory: rootDirectory
+    }
+  );
+
+  if (submoduleStatusResult.status !== 0) {
+    ensureGitMinimumVersion(gitPath);
+
+    throw new Error(
+      `git submodule status exited with status ${submoduleStatusResult.status}: ${submoduleStatusResult.stderr}`
+    );
+  }
+
+  // <hash> <path> (heads/main)\n
+  submoduleStatusResult.stdout.split('\n').forEach((line) => {
+    if (line) {
+      const [, submodulePath] = line.trim().split(' ');
+      submodulePaths.push(submodulePath);
+    }
+  });
+  return submodulePaths;
+}
+
+/**
+ * Augments the state value with modifications that are in git submodules
+ * @param rootDirectory - The root directory of the Git repository
+ * @param state - The current map of git path -> object hash. Will be mutated.
+ * @param gitPath - The path to the Git executable
+ * @internal
+ */
+export function applySubmodulesTreeState(
+  rootDirectory: string,
+  state: Map<string, string>,
+  gitPath?: string
+): void {
+  const submodulePaths: string[] = getSubmodulePaths(rootDirectory, gitPath);
+
+  if (submodulePaths.length === 0) {
+    return;
+  }
+
+  const submoduleLsTreeResult: child_process.SpawnSyncReturns<string> = Executable.spawnSync(
+    gitPath || 'git',
+    ['submodule', 'foreach', 'git', '--no-optional-locks', 'ls-tree', '-r', '--full-name', 'HEAD', '--'],
+    {
+      currentWorkingDirectory: rootDirectory
+    }
+  );
+
+  if (submoduleLsTreeResult.status !== 0) {
+    ensureGitMinimumVersion(gitPath);
+
+    throw new Error(
+      `git ls-tree for submodules exited with status ${submoduleLsTreeResult.status}: ${submoduleLsTreeResult.stderr}`
+    );
+  }
+
+  const stateBySubmodulePath: Record<string, Record<string, string>> = {};
+
+  let currentSubmodulePath: string = '';
+
+  // Entering '<submoduleFolder>'\n
```

Comment:
Is there a porcelain version of this command that's designed to be machine-readable?

Reply:
I think the `<mode> <type> <newhash>\t<path>\n` format is meant to be machine-readable.  The `\t` is very likely designed to avoid the need for an shell-specific escape when processing filesystem paths containing special characters (spaces, quotes, etc).

Reply:
I find the command usage from git official doc at https://git-scm.com/docs/git-submodule/2.19.0. The earliest version seems to be `2.19.0`. which meets the minimal version requirement of `2.20.0`